fdbclient/ClientDBInfo.cpp

```diff
@@ -0,0 +1,73 @@
+/*
+ * ClientDBInfo.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/ClientDBInfoRef.h"
+#include "fdbclient/MasterProxyInterface.h"
+
+void ClientDBInfoRef::addref() {
+	if (impl) {
+		reinterpret_cast<AsyncVar<ClientDBInfo>*>(impl)->addref();
+	}
+}
+
+void ClientDBInfoRef::delref() {
+	if (impl) {
+		reinterpret_cast<AsyncVar<ClientDBInfo>*>(impl)->delref();
+	}
+}
+
+ClientDBInfoRef::ClientDBInfoRef(void* impl) : impl(impl) {
+	addref();
+}
+ClientDBInfoRef::ClientDBInfoRef() {}
+
+ClientDBInfoRef::ClientDBInfoRef(const ClientDBInfoRef& other) : impl(other.impl) {
+	addref();
+}
+
+ClientDBInfoRef::ClientDBInfoRef(ClientDBInfoRef&& other) : impl(other.impl) {
+	other.impl = nullptr;
+}
+
+ClientDBInfoRef::~ClientDBInfoRef() {
+	delref();
+}
+
+ClientDBInfoRef& ClientDBInfoRef::operator=(const ClientDBInfoRef& other) {
+	delref();
```

Comment:
In the case of self-assignment, is this potentially going to destroy our object by calling delref first?

Reply:
Oh, good catch! Yes I think this is a bug